Rover: give sailboat motor sailing capability

IamPete1 opened the pull request:
This is quite a big change to sailboat, it gives functionality to use both sail and motor. This gives the sailboat functionality to act as a sailboat only or as traditional rover. This requires the addition of some new parameters that were previously shared. 

This adds two new aux functions, one to select the motoring state:
1: never motor
2: enable motoring, the code decides based on wind speed
3: always motor  

The second is a aux throttle, this is used in manual mode so 'main' throttle can be used for sail and the 'aux' throttle for motor. Or the other way with a options parameter. The aux throttle can also be used to limit the maximum auxiliary throttle. 

Requires https://github.com/ArduPilot/ardupilot/pull/11408 to test in SITL.

IamPete1 commented:
will need a patch for this https://github.com/ArduPilot/ardupilot/pull/11466 once it goes in, or the other way if this gets in first

rmackay9 commented on APMrover2/sailboat.cpp:
```diff
@@ -75,6 +75,42 @@ const AP_Param::GroupInfo Sailboat::var_info[] = {
+    // @Range: 0 5
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("MIN_WNDSPD", 10, Sailboat, sail_assist_windspeed, 0),
```
normally I prefer if we can put the MIN a the end of the parameter name..

IamPete1 replied:
fixed

rmackay9 commented:
So, I have to wonder if we should perhaps move the throttle control back to channel 3 (input and output) and use a new input for the mainsail.

peterbarker commented:
On Sun, 2 Jun 2019, Randy Mackay wrote:

> So, I have to wonder if we should perhaps move the throttle control back to channel 3 (input and output) and use a new input for the mainsail.

Probably should be using RCMAP?

Even better, my eliminate-rcmap PR and channel options :-/

IamPete1 commented:
This has now been tested, and seems to be working as expected. It also adds some sailboat navigation updates. A buffer angle has been added to so the heading controller is used for the 10deg bellow the no go angle, this results in much nice behaviour when the target heading is just on the no go angle, this happens on the last tack to a waypoint. A dead zone has been added for deciding which tack we are sailing on.

Bug fixed for acro tacking and auto mode tack time outs. 

I think the only remaining thing is to work out how to best handle the control of sail and motor in manual mode, this uses a aux channel but there may be a better way. Also it would be nice to add a waypoint type to change the motoring state, ie use only the sail for waypoint A then use the motor for waypoint B

rmackay9 commented on APMrover2/sailboat.h:
```diff
@@ -33,7 +34,7 @@ class Sailboat
 
     // update mainsail's desired angle based on wind speed and direction and desired speed (in m/s)
-    void  update_mainsail(float desired_speed);
+    float update_sail_control(float desired_speed, float throttle_out);
```
we should add a comment saying what is being returned..

IamPete1 replied:
fixed

rmackay9 commented on APMrover2/sailboat.h:
```diff
@@ -51,14 +52,30 @@ class Sailboat
+        ASSIST       = 1,
+        FORCE_MOTOR  = 2
+    };
+    uint8_t throttle_state; // throttle state used with throttle enum
```
if throttle state should always be one of the sailboat_throttle values then might as well make it's type "enum sailboat_throttle".  There's probably a better name for this enum though.  Often we use camel case for enums (although not always).  If it's going to use the underscore format then it should probably end with "_t" I think.

I would normally hide the set behind an accessor but that's probably not necessary and it eats some flash so it's probably ok.

IamPete1 replied:
fixed

rmackay9 commented on APMrover2/sailboat.h:
```diff
@@ -68,14 +85,28 @@ class Sailboat
     AP_Float sail_heel_angle_max;
     AP_Float sail_no_go;
+    AP_Float max_cross_track;
+    AP_Float loit_radius;
```
is the extra loiter radius really necessary?  I guess the point is that a user might want to have different radius depending upon whether they are using a motor or are relying on just the sail.. it seems a bit of a luxury to me..

IamPete1 replied:
Yeah that was the idea, we can use the normal loiter rad for the motor loiter and the new sailboat one for a sailing loiter. Not much point in using the motor in a loiter if you still have to have the big radius you need for the sailing loiter. Now we have a enable param I feel less bad about adding in more sailing stuff.

rmackay9 commented on APMrover2/sailboat.cpp:
```diff
@@ -127,111 +171,178 @@ void Sailboat::update_mainsail(float desired_speed)
+    }
+
+    // see if we should limit the throttle allowed
+    if ( (sail_options & options::AUX_THROTTLE) != 0 && (sail_options & options::AUX_THROTTLE_LIMIT) != 0 ) {
```
minor formatting issue - a little extra space between the brackets on this line

rmackay9 commented on APMrover2/sailboat.cpp:
```diff
@@ -127,111 +171,178 @@ void Sailboat::update_mainsail(float desired_speed)
+
+    // see if we should limit the throttle allowed
+    if ( (sail_options & options::AUX_THROTTLE) != 0 && (sail_options & options::AUX_THROTTLE_LIMIT) != 0 ) {
+        RC_Channel *rc_ptr = rc().find_channel_for_option(RC_Channel::AUX_FUNC::SAILBOAT_THR);
```
I'm a little concerned that this breaks the pattern we have in Copter and Rover re consumption of the pilot input.  We normally consume the pilot's input at the top of the flight mode code.  This unfortunately buries is two levels below the flight code.  When we break this pattern it makes it much more difficult to figure out what is going on.

rmackay9 replied:
I think we should probably use the regular throttle channel for throttle and then allow the pilot to control the main sail with channel 4.

IamPete1 replied:
I have removed this throttle limit stuff, not sure how usefully it would be in anycase

rmackay9 commented on APMrover2/sailboat.h:
```diff
@@ -24,7 +24,8 @@ class Sailboat
     // enabled
-    bool enabled() const { return enable != 0;}
+    bool sail_enabled() const { return enable != 0;}
+    bool nav_enabled();
```
could we add a comment above "nav_enabled()" to clarify?  does it actually mean "tacking_enabled" perhaps?

IamPete1 replied:
added a small description, currently this is just if we are tacking upwind, it may become more if/when i get this drag polar stuff working

rmackay9 commented on APMrover2/sailboat.cpp:
```diff
@@ -90,25 +128,31 @@ Sailboat::Sailboat()
-        rover.g2.wp_nav.set_default_overshoot(10);
+    }
+
+    // if we have a throttle of some sort allow to use it
```
it's a crazy edge case but it could also theoretically be an omni sailboat.. insanely unlikely though..

IamPete1 replied:
fixed, would love to see a onmi sailboat...

IamPete1 commented:
I have removed a the complicated throttle stuff, this now has a aux option that can be used to control the only sail in  only manual mode, if it is not set then it will be on the throttle as usual. This removes the need for the complicated arming checks. I have also added a fix for the PID issue that is on master following the recent changes.

IamPete1 commented:
I have done a mavlink PR to add the waypoint type here: https://github.com/ArduPilot/mavlink/pull/103

and a quick demo vid in SITL https://youtu.be/ho2hkkajMBo

This is now complete I think

IamPete1 commented:
added forgotten case for reading back waypoints

rmackay9 commented on libraries/APM_Control/AR_AttitudeControl.cpp:
```diff
@@ -576,15 +576,15 @@ float AR_AttitudeControl::get_sail_out_from_heel(float desired_heel, float dt)
-    // constrain p to only be positive
-    if (!is_positive(p)) {
+    // constrain p to only be negative
+    if (!is_negative(p)) {
```
So the P and I terms of the HEEL PID will be negative.  That cannot be correct.  @lthall maybe we need to discuss the fix after the latest changes?

IamPete1 replied:
Leonard helped my work this one out, my original error was calculated wrong, the correct method now used gives -1 * the error I was using before. Hence the change from is_positive to is_negative. Then the whole lot is multiplied by -1 at the end so it looks the same to everything else. I have tested this on the boat (on the stand).

IamPete1 commented:
rebased, but waiting on https://github.com/ArduPilot/mavlink/pull/103

rmackay9 commented:
To help things along I've got a few commits at the end of [this branch](https://github.com/rmackay9/rmackay9-ardupilot/commits/iampete1-motor-sail) which is functionally the same (I think) but makes the rc input handling a little more consistent with the existing rover code.

Perhaps one of the last things I'd like to have changed before we merge is how the speed controller is called from within the Mode:;calc_throttle and Sailboat::update_sail_control() functions.  I'm not keen on update_sail_control() relaxing the throttle controler.  Essentially we are running the speed controller in calc_throttle and then throwing away the results in update_sail_control.  To try and keep things understandable to future programmers I think we need to consolidate the poking/proding of external objects like the speed controller all in one place.  My guess is that Sailboat's update_sail_control function should do the whole job job of deciding what the mainsail and throttle outputs should be to achieve a given speed and calc_throttle will call this new function if we are a sailboat.

IamPete1 commented:
@rmackay9 I have cherry picked over your fixes, Thanks!. I have had a go at refactoring calc_throttle, if your happy I can squash abit.

IamPete1 commented:
save sailing speed control for later PR and only call throttle controller if we use the value

rmackay9 commented on APMrover2/sailboat.cpp:
```diff
@@ -167,24 +167,16 @@ void Sailboat::get_pilot_desired_mainsail(float &mainsail_out)
     if (throttle_state_t == Sailboat_Throttle::FORCE_MOTOR) {
         rover.g2.motors.set_mainsail(100.0f);
-        return throttle_out;
+        return rover.control_mode->get_throttle_out(desired_speed);
```
Hi, not a big fan of how the mode class is calling Sailboat's update_sail_control which is then calling back into Mode::get_throttle_out().  That's very confusing for the next developer

IamPete1 replied:
I guess we could have a  bool function to say if we should call get_throttle_out then call it from within calc throttle, like how the heading stuff works, there is on_indirect_route() to decide if we should call calc_heading()

rmackay9 commented:
I've created a simplified version of the main motor assistance feature of this PR in this branch: https://github.com/rmackay9/rmackay9-ardupilot/commits/iampete1-motor-sail3

I'm sure it won't cause huge joy that it's so chopped down but I wanted to reduce it to a smaller set of changes so that it's easier to review, test and merge.  Other changes can go in later as smaller PRs.

I haven't tested this in SITL yet but I hope to tomorrow

rmackay9 commented:
@IamPete1, I've created this alternative PR: https://github.com/ArduPilot/ardupilot/pull/12107 which has a reduced set of features from this PR I think it's merge-able (assuming you don't find any bugs).

I hope you don't mind too much that I reworked it a bit.  Most of the changes I made were non-functional changes (i.e. renamed variables and methods and moved things around) but I think I may have also fixed an issue with how tacking was enabled - for this I split the "throttle_assist" function into two functions one for assist-during-tacking the other assist-during-low-wind.

Anyway, please tell me if you see any issues with the new PR.  We can work on getting the other bits like changes to the tacking logic in as a follow-up. txs!

IamPete1 commented:
with https://github.com/ArduPilot/ardupilot/pull/12774 all of this is either in already or is in open PR's in smaller chunks.
